In [1]:
p = 2  # p and q should be prime numbers (mostly large)
q = 7

In [2]:
n = p * q

In [4]:
phi = (p-1) * (q-1)

In [5]:
print(n,phi)

14 6


In [6]:
# First find 'e' such that e and phi are co-primes
# - being co-prime means that their GCD will be 1 OR
# No shared factors except 1

In [8]:
def gcd(a,b):      # finding GCD of two numbers
    while b!= 0:
        a,b = b, a % b
    return a

In [9]:
def get_e(phi):
    e = 2          # start with 2 bcz 1 is divisible for all nums
    
    while True:
        if gcd(e,phi) == 1:   # it means e and phi are co-prime
            break      # we have a good 'e'
        
        e += 1      # try next number
    
    return e

In [11]:
e = get_e(phi)
print(e)

5


In [12]:
# find 'd' such that ed mod phi == 1

In [19]:
def get_d(init_val = 1):
    d = init_val   
    
    while True:
        if (e * d % phi) == 1:
            break   # we have a good 'd'
        
        d += 1
        
    return d

In [21]:
d = get_d(10)     # start with 10 so that 11 comes match condition
print(d)

11


In [22]:
# let's encrypt just one character
msg = 2

In [27]:
# we're going to publish  e and n all over the world

enc = msg**e % n  # Alice do and sent it to Bob on the internet

In [28]:
print(enc)

4


In [29]:
# let's decrypt msg to original
decrypt = enc**d % n

In [30]:
print(decrypt)

2


In [31]:
# The Core Concept: 'e' and 'n are well known' -- but 'd' is hidden

# let's say, I wanted to tell you the amount I owe you but in a way that provides a guarntee that only I could have written that amount -- sort like a digital signature

In [32]:
amount = 1000

In [33]:
  # 'sign' it -- meaning encrypt it with 'd' this time instead of 'e'

In [35]:
p = 199
q = 131
n = p * q
phi = (p-1) * (q-1)
e = get_e(phi)
d = get_d()
print('n  :  ' , n)

print('e  :  ' , e)
print('d  :  ' , d)
print('phi :  ' , phi)


n  :   26069
e  :   7
d  :   22063
phi :   25740


In [36]:
sign  = amount **d % n   # encrypt with d
print(sign)

17403


In [37]:
# Now, since e is public key, anyone can verify that I've signed
dec = sign**e % n
print(dec)

1000


# so anyone can crack 1000 it means it is not hidden, 1000 will be made if and only if 17403 is encrypted by my public key d

In [38]:
# so d is corresponding to e and only i've e it means i encrypt it

# point: if something is cracked by d it means that thing is
# encrypted by e and vice versa

# you don't actually sign the full msg.you take the hash of msg and sign that.

In [39]:
p = 983
q = 719
n = p * q
phi = (p-1) * (q-1)
e = get_e(phi)
d = get_d()
print('n  :  ' , n)

print('e  :  ' , e)
print('d  :  ' , d)
print('phi :  ' , phi)

n  :   706777
e  :   3
d  :   470051
phi :   705076


In [40]:
def bad_hash(msg):
    s = 0;
    for c in msg:
        s += ord(c)
    
    return int(s % 1e10)

In [41]:
message = "I owe you a total of PKs 1000 Collect on April 28, 2023" 

In [43]:
digest = bad_hash(message) # poor hash function just undersatnd the functionality of hash function

In [46]:
digest

4242

In [48]:
sign = digest ** d % n  # now we sign it OR encrypt the digest
print(sign)

565054


In [49]:
# send this to the public

(message , sign)

('I owe you a total of PKs 1000 Collect on April 28, 2023', 565054)

In [50]:
# public identifies that it was you or not!!!

digest = bad_hash(message)
digest

4242

In [51]:
dec = sign ** e % n
print(dec)

4242


In [53]:
dec == digest    # All the set!
                # value is same it means that is your sign bcoz the corresponding value of d
                # is only you know 

True